lib/fh_cluster.js

```diff
@@ -62,12 +64,163 @@ module.exports = function fhCluster(workerFunc, optionalNumWorkers, optionalBack
     throw invalidWorkerTask.error;
   }
 
-  setOnWorkerListeningHandler();
-  setupOnExitBackOff(backoffStrategy, backoffResetTimeout);
-
-  start(workerFunc, numWorkers);
+  if (cluster.isMaster) {
+    startMaster( numWorkers, backoffStrategy, backoffResetTimeout );
+  } else {
+    startWorker( workerFunc, gracefullShutdownFn );
+  }
 };
 
+/********************************************/
+/* MASTER - code that powers master process */
+/********************************************/
+
+function startMaster(numWorkers, backoffStrategy, backoffResetTimeout) {
+  //Need to store the number of workers
+  g_numWorkers = numWorkers;
+
+  // set up master process event handlers
+  cluster.on('exit', setupOnExitBackOff(backoffStrategy, backoffResetTimeout));
+  cluster.on('exit', setupBoundsTasks);
+  cluster.on('listening', setOnWorkerListeningHandler);
+  process.on('SIGTERM', cleanMasterShutdown('SIGTERM'));
+  process.on('SIGHUP', cleanMasterShutdown('SIGHUP'));
+  process.on('SIGINT', cleanMasterShutdown('SIGINT'));
+
+  // start workers
+  _.times(numWorkers, function() {
+    cluster.fork();
+  });
+}
+
+function setupOnExitBackOff(strategy, backoffResetTimeout) {
+  return function(worker) {
+    if (worker.suicide) {
+      log.debug('Worker #', worker.id, 'commited suicide, won\'t retry');
+      return;
+    }
+    var nextRetry = strategy.next();
+    log.info('Worker #', worker.id, 'exited. Will retry in',
+      nextRetry, 'ms');
+
+    setTimeout(function() {
+      cluster.fork();
+      resetBackoffResetTimeout(strategy, backoffResetTimeout);
+    }, nextRetry);
+  };
+}
+
+function setOnWorkerListeningHandler(worker, address) {
+  var host = address.address || address.addressType === 4 ? '0.0.0.0' : '::';
+  var addr = host + ':' + address.port;
+  log.info('Cluster worker',  worker.id, 'is now listening at', addr);
+
+  //A new process is listening, need to check if any bound tasks need to be assigned.
+  boundTasks.setUpBoundTasks(singleWorkerTasks, cluster);
+}
+
+function setupBoundsTasks(){
+  //If a worker exits, check that all bound workers are on a running worker.
+  boundTasks.setUpBoundTasks(singleWorkerTasks, cluster);
+}
+
+function resetBackoffResetTimeout(backoffStrategy, timeout) {
+  if (timeout) {
+    clearTimeout(timeout);
+  }
+
+  return setTimeout(function() {
+    backoffStrategy.reset();
+  }, 60*60*1000);
+}
+
+function cleanMasterShutdown(type) {
+  return function (cb) {
```

Comment:
For this sonarqube violation, it's looking for a semi-colon at line 148. I think the severity of that rule has been increased since this ran, so please add that also! :)